fairseq2n/src/fairseq2n/data/data_pipeline.cc

```diff
@@ -224,6 +225,39 @@ data_pipeline::round_robin(std::vector<data_pipeline> pipelines)
     return data_pipeline_builder{std::move(factory)};
 }
 
+data_pipeline_builder
+data_pipeline::sample(
+    std::vector<data_pipeline> pipelines,
+    std::optional<std::vector<float>> weights)
+{
+    if (pipelines.empty())
+        throw_<std::invalid_argument>(
+            "`pipelines` does not contain any elements. Can not sample from empty set.");
+
+    bool is_broken = std::any_of(
+        pipelines.begin(), pipelines.end(), [](const data_pipeline &pipeline)
+        {
+            return pipeline.is_broken();
+        });
+    if (is_broken)
+        throw_<std::invalid_argument>(
+            "At least one of the specified data pipelines is broken and cannot be used in sample.");
+
+    if (!weights)
+        weights = std::vector<float>(pipelines.size(), 1.0F / static_cast<float>(pipelines.size()));
+    else if (weights.value().size() != pipelines.size())
+        throw_<std::invalid_argument>(
+            "`weights` should have the same size as `pipelines`.");
```

Comment:
Nit: Can you also include their sizes in the error message. You can check out other op functions as an example. Note that `throw_<>()` handles message formatting, so it should be relatively easy.